File: wingmon.cpp
#include <cstring>
#include <iostream>

#include "WingConsole.h"

using namespace std;

int
main()
{
    printf("Discovering Behringer Wing consoles...\n");
    auto discovered = WingConsole::scan();
    
    if (discovered.empty()) {
        printf("No Behringer Wing consoles found.\n");
        return 1;
    } else {
        printf("Found %zu consoles(s):\n", discovered.size());
        for (size_t i = 0; i < discovered.size(); i++) {
            printf("    %zu. %s (%s)\n", i+1, discovered[i].name.c_str(), discovered[i].ip.c_str());
        }
        fflush(stdout);
    }

    printf("Connecting to Behringer Wing console %s\n", discovered[0].name.c_str());

    auto console = WingConsole::connect(discovered[0].ip);
    console.onNodeData = [&](auto id, auto data) {
        string name = NodeDefinition::nodeIdToName(id);
        if (name.empty()) {
            printf("<UnknownId:0x%08x> = %s\n", id, data.getString().c_str());
        } else {
            printf("%s = %s\n", name.c_str(), data.getString().c_str());
        }
    };

    cout << "Monitoring for changes..." << endl;
    console.read();
    cout << "... device disconnected." << endl;

    return 0;
}

File: tools/wingmon.rs
use std::io::{self, Write};
use libwing::{WingConsole, WingResponse};
use std::result::Result;

fn main() -> Result<(),libwing::Error> {
    // Discover Wing devices
    let devices = WingConsole::scan(true)?;
    if devices.is_empty() {
        eprintln!("No Wing devices found!");
        return Ok(());
    }

    // Print discovered devices
    println!("Found {} Wing device(s):", devices.len());
    for (i, dev) in devices.iter().enumerate() {
        println!("{}. {} at {} (Model: {}, Firmware: {})",
            i + 1, dev.name, dev.ip, dev.model, dev.firmware);
    }

    // Let user choose if multiple devices found
    let device = if devices.len() > 1 {
        print!("Select device (1-{}): ", devices.len());
        io::stdout().flush()?;
        let mut input = String::new();
        io::stdin().read_line(&mut input)?;
        let selection = input.trim().parse::<usize>().unwrap_or(0);
        if selection < 1 || selection > devices.len() {
            eprintln!("Invalid selection!");
            return Ok(());
        }
        &devices[selection - 1]
    } else {
        &devices[0]
    };

    println!("Connecting to {} at {}...", device.name, device.ip);
    let mut console = WingConsole::connect(&device.ip)?;
    println!("Connected!");

    loop {
        if let WingResponse::NodeData(id, data) =  console.read()? {
            println!("{} {} = {}",
                id,
                WingConsole::id_to_name(id).unwrap_or(&format!("<Unknown:{}>", id)),
                data.get_string());
        }
    }
}
